[E2E] [QIT] Add Jetpack connection support for QIT E2E tests

mgascam opened the pull request:
Fixes #WOOPMNT-5336

#### Changes proposed in this Pull Request

This is a stacked PR that builds on top of the [basic QIT E2E](https://github.com/Automattic/woocommerce-payments/pull/11047) testing functionality to address the key challenge of establishing a Jetpack connection for WooPayments in test environments.

The previous foundation established basic E2E testing infrastructure, but lacked the ability to connect WooPayments to a real test account via Jetpack. This PR solves that by providing automated Jetpack connection setup for already onboarded test accounts.

**Key features:**
1. **WP-CLI QIT Development Commands**: New `WP_CLI_QIT_Dev_Command` class providing:
   - `wp woopayments qit_jetpack_connection` - Configures Jetpack connection for pre-onboarded test sites
   - `wp woopayments qit_jetpack_status` - Shows connection status for debugging
   - Automatic test mode configuration and account data refresh from server

2. **Enhanced E2E Test Runner**: Updated `e2e-runner.sh` with:
   - Plugin build caching based on source file signatures (avoids unnecessary rebuilds)
   - Jetpack credential passing via environment variables

3. **Connection Validation Tests**:
   - Tests to verify WooPayments shows configuration options (not Connect screen)
   - Validation that account data is properly fetched from the server

4. **Configuration Management**:
    - Custom volume mounts in qit.yml for cleaner bootstrap file paths
    - Bootstrap scripts use wp eval-file to execute PHP setup code in QIT container
    - Environment variable handling for site ID, blog token, and user token

**Why This Approach:**
The added WP-CLI commands do essentially the same thing as the WCPay Dev Tools plugin we use for current E2E tests. Initially I tried using the dev plugin directly, but downloading and installing it in the QIT environment proved tricky. This workaround was inspired by @dmvrtx's work in https://github.com/Automattic/woocommerce-payments/pull/10811. However, initializing a test drive account turned out to be complex, so I opted to use an already onboarded account instead. This follows the same approach as our current E2E tests, adapted for QIT's needs. The WP-CLI commands are loaded via`wp eval-file` in the QIT container and never ship to production, keeping test infrastructure completely isolated from production code. **Recent Refactoring**: The bootstrap setup has been refactored for better clarity and maintainability:

- Custom volume mount (`./e2e/bootstrap:/qit/bootstrap:ro`) provides cleaner paths in setup scripts
- Renamed files to be more descriptive: `qit-jetpack-connection.php` and `qit-jetpack-status.php`
- CLI command names updated to `qit_jetpack_connection` and `qit_jetpack_status` to clearly indicate their QIT-specific purpose
- Added documentation comments explaining the volume mount setup throughout


<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

**Prerequisites:**
- QIT CLI installed and configured (ie. run `composer install`)
- Access to a pre-onboarded WooPayments test account with Jetpack credentials. The following steps worked well for me:
  1. Spin a fresh JN site with WooCommerce, WooPayments and Jetpack Debug Tools 
  2. Follow the minimum steps to onboard with a test account
  3. Go to Jetpack Debug > Broken Token to find the current tokens and blog ID
  4. Copy `tests/qit/config/default.env` to `tests/qit/config/local.env`
  5. Fill in E2E credentials from your onboarded test account:
```
E2E_JP_SITE_ID=your_site_id_here
E2E_JP_BLOG_TOKEN=your_blog_token_here
E2E_JP_USER_TOKEN=your_user_token_here
``` 
<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->
**Running the tests:**
1. Run E2E tests: `npm run test:qit-e2e`
2. Connection validation tests should pass

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

mgascam committed:
Add QIT development tools and E2E testing setup

- Introduced WP-CLI commands for QIT setup and status checks to assist in development environments.
- Created a new class `WP_CLI_QIT_Dev_Command` for handling Jetpack connection setup and account data refresh.
- Added E2E test runner script and configuration for QIT, including basic tests for WooCommerce Payments functionality.
- Updated package.json to include a new script for running QIT E2E tests.
- Enhanced default environment configuration for QIT testing with necessary credentials.
- Implemented setup script for WooCommerce Payments E2E tests to configure products, customers, and Jetpack connections.
- Added ESLint configuration for E2E tests to accommodate Playwright testing framework.
- Created initial E2E test cases to verify basic functionality and plugin activation.

mgascam committed:
Refactor QIT development tools for WooPayments: update comments, enhance setup functions, and add connection validation tests

mgascam committed:
Remove QIT development tools and simplify E2E test setup for WooCommerce Payments

mgascam committed:
Add E2E test credentials and enhance WooPayments setup scripts

mgascam committed:
Add changelog

mgascam committed:
Amend comment

mgascam committed:
Merge branch 'dev/qit-e2e-foundation' into dev/qit-e2e-basic-checkout

mgascam committed:
Merge branch 'develop' into dev/qit-e2e-foundation

mgascam committed:
Merge branch 'dev/qit-e2e-foundation' into dev/qit-e2e-basic-checkout

mgascam committed:
Revert changes to composer.*

mgascam committed:
Upgrade qit cli package

mgascam committed:
Merge branch 'dev/qit-e2e-foundation' into dev/qit-e2e-basic-checkout

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `11049` or branch name `dev/qit-e2e-basic-checkout` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=dev/qit-e2e-basic-checkout) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 9b494e61ae08488086b93dee38557176433b16c9
- Build time: 2025-11-10 17:11:29 UTC

_Note: the build is updated when a new commit is pushed to this PR._

mgascam committed:
Add 'tests/qit/e2e' to path ignore patterns in Jest config

mgascam committed:
Merge branch 'dev/qit-e2e-foundation' into dev/qit-e2e-basic-checkout

mgascam committed:
Merge branch 'develop' into dev/qit-e2e-foundation

mgascam committed:
Merge branch 'dev/qit-e2e-foundation' into dev/qit-e2e-basic-checkout

mgascam committed:
Add wp-cli-stubs dependency and update psalm-loader for static analysis

rtio reviewed (commented):
I got 2 tests that failed.

QIT report: https://qit.woo.com/?qit_results=4886709.kMG2Aaw8PtWU586tsbSYa5z2BFMtuzpBuOLdQYmHhTDrYHMmTYX6KMbvlH5CIoP2

test store: https://spoonbill-of-roadrunners.jurassic.ninja/
creds 391ec-pb/
console output 391ed-pb/

```
2 failed
    [test] Payments (Run) › woocommerce-payments/local/basic.spec.js:54:5 › Verify WooCommerce Payments plugin activation
    [test] Payments (Run) › woocommerce-payments/local/checkout.spec.js:15:2 › WooPayments Connection Status › should verify WooPayments is connected (not showing Connect screen)
```

mgascam committed:
Merge branch 'develop' into dev/qit-e2e-foundation

mgascam committed:
Merge branch 'dev/qit-e2e-foundation' into dev/qit-e2e-basic-checkout

mgascam committed:
Refine plugin row locator to exclude update rows in activation test

mgascam committed:
Merge branch 'dev/qit-e2e-foundation' into dev/qit-e2e-basic-checkout

mgascam commented:
Thanks for the review @rtio!

> I got 2 tests that failed.
> 
> QIT report: https://qit.woo.com/?qit_results=4886709.kMG2Aaw8PtWU586tsbSYa5z2BFMtuzpBuOLdQYmHhTDrYHMmTYX6KMbvlH5CIoP2
> 
> test store: https://spoonbill-of-roadrunners.jurassic.ninja/ creds 391ec-pb/ console output 391ed-pb/
> 
> ```
> 2 failed
>     [test] Payments (Run) › woocommerce-payments/local/basic.spec.js:54:5 › Verify WooCommerce Payments plugin activation
>     [test] Payments (Run) › woocommerce-payments/local/checkout.spec.js:15:2 › WooPayments Connection Status › should verify WooPayments is connected (not showing Connect screen)
> ```

I noticed the `Verify WooCommerce Payments plugin activation` was also failing on the base branch and fixed it and also rebased this branch. I tried a new test run and it passes for me: https://qit.woo.com/?qit_results=4890413.NhswLJdMaModEaCjbNqRu62nkoefL3IU4sIvQ1U7n8wuq1lBX78nhTAIV7395I8A.

Maybe there's something with this specific test site 🤔 ? I'm taking a look, thanks for sharing your credentials!

mgascam commented:
I tried a new run, using the https://spoonbill-of-roadrunners.jurassic.ninja/  tokens and it passed this time. https://qit.woo.com/?qit_results=4890881.GrnxJ6y846wUZxGIeesBsTvvVWRTJrXfGfTLdXmoNlcWUtvsEbBG05T805wRGS6A

mgascam committed:
Rename misspelled vars on env file

mgascam committed:
Merge branch 'develop' into dev/qit-e2e-foundation

mgascam committed:
Merge branch 'dev/qit-e2e-foundation' into dev/qit-e2e-basic-checkout

mgascam committed:
Merge branch 'develop' into dev/qit-e2e-foundation

mgascam committed:
Merge branch 'dev/qit-e2e-foundation' into dev/qit-e2e-basic-checkout

mgascam committed:
WIP refactor cli commands

mgascam committed:
Refactor QIT Jetpack connection setup and status scripts

mgascam committed:
Remove wp-cli-stubs dependency from composer.json and psalm-loader.php

mgascam committed:
Add safety checks for JetPack commands and update environment setup script

mgascam committed:
Merge branch 'develop' into dev/qit-e2e-foundation

mgascam committed:
Merge branch 'develop' into dev/qit-e2e-foundation

mgascam committed:
Merge branch 'dev/qit-e2e-foundation' into dev/qit-e2e-basic-checkout

mgascam committed:
Merge remote-tracking branch 'origin/develop' into dev/qit-e2e-basic-checkout

mgascam committed:
Add changelog

mgascam committed:
Merge branch 'develop' into dev/qit-e2e-basic-checkout
